### DISCUSSION TITLE
Composite body connecting two rigid bodies

### DISCUSSION THREAD

**Discussion by curtiscjohnson** [google-deepmind/mujoco#52] [Asking for Help] [answered]
I'm new to Mujoco and there are a lot of features I am interested in using. I am trying to model a somewhat unique robot arm which has pneumatically actuated joints which connect rigid links. [This video](https://www.youtube.com/watch?v=WMoCoudksFc) is what the robot looks like moving around. The orange parts are the flexible pneumatic joints.

I've been trying to think of ways to model this robot with Mujoco and so far I haven't been able to figure out how to get this to work. Here are the things I've thought of and what issues I've had trying to implement them:

- Model the pneumatic joints as a composite cylinder. It is unclear to me how I can than "actuate" the cylinder the way it does in the video above. I'm also not sure how to "weld" the composite cylinder to the next rigid body.
- Use the kinematic linkage feature in a composite rope. But I also am not sure how that could be actuated or connected to the next rigid body either. 
- I would like to use the actuator dynamics feature of Mujoco's actuator model, so I thought of using pneumatic cylinders but those don't act like a flexible body the way these joints do. 

Does anyone more experienced with Mujoco have any suggestions of what I can try? Or is this just not something that Mujoco can handle? I can't really find any suggestions online for this type of system. 

Thanks!!

**Comment by yuvaltassa** [comment, marked as answer]
Hi,

As mentioned in [this issue](https://github.com/deepmind/mujoco/issues/43) composite objects are just macros. If you load and save an XML with a composite body (e.g. in `simulate`), the macro will be expanded and you will be able to modify the details of your model as you please.

Pneumatic cylinder actuators capture the fact that air pressure is an internal state of the actuator and are unrelated to the kinematics.

**Reply by idoyechimo** [reply]
Hello Yuval,
I am trying to follow your instructions in order to get access to a specific body in a composite rope. but whenever I try pressing the "Save XML" button in "simulate", nothing happens.
any idea on what can be done?

**Reply by yuvaltassa** [reply]
Look in the directory where your `simulate` binary lives.
Can you not see a file called `mjmodel.xml` ?

**Reply by idoyechimo** [reply]
Found it. Thank you.
